src/protocol/decoder.rs

```diff
@@ -0,0 +1,412 @@
+use crate::{
+    models::ERC20Token,
+    protocol::{
+        errors::InvalidSnapshotError,
+        models::{BlockUpdate, ProtocolComponent, TryFromWithBlock},
+        state::ProtocolSim,
+    },
+};
+use std::{
+    collections::{hash_map::Entry, HashMap},
+    future::Future,
+    pin::Pin,
+    str::FromStr,
+    sync::Arc,
+};
+use thiserror::Error;
+use tokio::sync::RwLock;
+use tracing::{debug, warn};
+use tycho_client::feed::{synchronizer::ComponentWithState, FeedMessage, Header};
+use tycho_core::Bytes;
+
+#[derive(Error, Debug)]
+pub enum StreamDecodeError {
+    #[error("{0}")]
+    Fatal(String),
+}
+
+#[derive(Default)]
+struct DecoderState {
+    tokens: HashMap<Bytes, ERC20Token>,
+    states: HashMap<String, Box<dyn ProtocolSim>>,
```

Comment:
Why String and not Bytes?

Reply:
Because component ids are strings not bytes.